### PR TITLE
Extensibility question - OnDoUpdateItemVariantBeforeModify and OnDoUpdateItemVariantBeforeModifyand

### PR DESCRIPTION
**Pull request by viktorglemne**
New events as described in issue #25377, OnDoUpdateItemVariantBeforeModify and OnDoUpdateItemVariantBeforeModifyand.<br /><br />Internal work item: AB#494617

### CODE REVIEW

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al [hunk 1]
```diff
@@ -347,4 +347,21 @@ codeunit 30177 "Shpfy Product Events"
...
     begin
     end;
-}
\ No newline at end of file
+
+    [IntegrationEvent(false, false)]
+    /// <summary> 
+    /// Raised Before Modify on Items.
+    /// </summary>
+    internal procedure OnDoUpdateItemVariantBeforeModify(Shop: Record "Shpfy Shop"; var ShopifyProduct: Record "Shpfy Product"; ShopifyVariant: var Item: Record Item; var IsModified: Boolean);
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al:355]: I think we should stick to similiar names like the other events, that use `ShopifyShop` for `Record "Shpfy Shop"`.

Additionaly this event looks a little bit wrong. The `ShopifyVariant:` should be removed from the signature

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al [hunk 2]
```diff
@@ -97,11 +99,17 @@ codeunit 30188 "Shpfy Update Item"
...
+        IsModified: Boolean;
     begin
         if ItemVariant.Description <> ShopifyVariant.Title then begin
             ItemVariant.Description := ShopifyVariant.Title;
-            exit(ItemVariant.Modify());
+            IsModified := true;
         end;
+        ProductEvents.OnDoUpdateItemVariantBeforeModify(Shop, ShopifyVariant, ItemVariant, IsModified);
+        if IsModified then
+            exit(item.Modify());
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al:111]: A small Copy/Paste Issue
```suggestion
            exit(ItemVariant.Modify());
```

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al [hunk 3]
```diff
@@ -347,4 +347,21 @@ codeunit 30177 "Shpfy Product Events"
     internal procedure OnBeforeSendUpdateShopifyProductVariant(ShopifyShop: Record "Shpfy Shop"; var ShopifyVariant: Record "Shpfy Variant"; xShopifyVariant: Record "Shpfy Variant")
     begin
     end;
-}
\ No newline at end of file
+
+    [IntegrationEvent(false, false)]
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al:351]: All new Events should include an xml documentation for any parameter.

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al (hunk 1 above)
Comment (viktorglemne) [reply Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al]: I agree fully on your suggestions! I have changed as according and committed the changes.

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al (hunk 3 above)
Comment (viktorglemne) [reply Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al:351]: I agree fully on your suggestions! I have changed as according and committed the changes.

**Comment by viktorglemne** [comment]
@microsoft-github-policy-service agree company="Navcite"

**Review by AndreiPanko — DISMISSED** [review]

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al [hunk 4]
```diff
@@ -347,4 +347,29 @@ codeunit 30177 "Shpfy Product Events"
...
+    /// <summary> 
+    /// Raised Before Modify on Items.
+    /// </summary>
+    /// <param name="ShopifyShop">Parameter of type Record "Shopify Shop".</param>
+    /// <param name="ShopifyProduct">Parameter of type Record "Shpfy Product".</param>
+    /// <param name="Item">Parameter of type Record "Item".</param>
+    /// <param name="IsModified">Parameter of type Boolean.</param>
+    internal procedure OnDoUpdateItemBeforeModify(var ShopifyShop: Record "Shpfy Shop"; var ShopifyProduct: Record "Shpfy Product"; var Item: Record Item; IsModified: Boolean);
```
Comment (pri-kise) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al:359]: @viktorglemne I think you forgot to set the parameter `IsModified` as `var`, that this parameter can be updated inside an event.
@AndreiPanko

Comment (viktorglemne) [reply Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyProductEvents.Codeunit.al]: Yes that was an miss! @pri-kise its now updated

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al [hunk 5]
```diff
@@ -97,11 +99,17 @@ codeunit 30188 "Shpfy Update Item"
...
     begin
         if ItemVariant.Description <> ShopifyVariant.Title then begin
             ItemVariant.Description := ShopifyVariant.Title;
-            exit(ItemVariant.Modify());
+            IsModified := true;
         end;
+        ProductEvents.OnDoUpdateItemVariantBeforeModify(Shop, ShopifyVariant, ItemVariant, IsModified);
+        if IsModified then
+            exit(ItemVariant.Modify());
+        exit(false);
```
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al:112]: I am not the biggest fan of this change. This event can override IsModified, and don't sync the item even though it has changed.

I would prefer something like this
 
ProductEvents.OnDoUpdateItemBeforeModify(Shop, ShopifyProduct, Item, IsModifiedByEvent);
        if IsModified or IsModifiedByEvent then
            exit(Item.Modify());
        exit(false);

File: Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al [hunk 6]
```diff
@@ -86,8 +86,10 @@ codeunit 30188 "Shpfy Update Item"
                     Item."Vendor No." := Vendor."No.";
                 end;
         end;
+        ProductEvents.OnDoUpdateItemBeforeModify(Shop, ShopifyProduct, Item, IsModified);
```
Comment (onbuyuka) [inline Apps/W1/Shopify/app/src/Products/Codeunits/ShpfyUpdateItem.Codeunit.al:89]: Same concern here, I would prefer another boolean than IsModified

**Review by onbuyuka — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Thanks for reporting this. We agree, and we’ll publish a fix asap, either in an update for the current version or in the next major release. Please do not reply to this, as we do not monitor closed issues. If you have follow-up questions or requests, please create a new issue where you reference this one.

Build ID: 14787.

**Review by JesperSchulz — APPROVED** [review]

**Comment by JesperSchulz** [comment]
Availability update: We will publish a fix for this issue in the next update for release 23.

Build ID to track: 14841.

**Pull request merged**